packages/mongodb-ts-autocomplete/src/index.ts

```diff
@@ -0,0 +1,215 @@
+  private readonly context: AutocompletionContext;
+  private connectionSchemas: Record<string, ConnectionSchema>;
+  private readonly autocompleter: Autocompleter;
+
+  constructor({ context, autocompleterOptions }: MongoDBAutocompleterOptions) {
+    this.context = context;
+    this.autocompleter = new Autocompleter(autocompleterOptions);
+
+    this.connectionSchemas = Object.create(null);
+    this.connectionSchemas;
+
+    this.autocompleter.updateCode(autocompleteTypes);
+  }
+
+  addConnection(connectionId: string): ConnectionSchema {
+    if (!this.connectionSchemas[connectionId]) {
+      this.connectionSchemas[connectionId] = new ConnectionSchema();
+    }
+    return this.connectionSchemas[connectionId];
+  }
+
+  getConnectionCode(connectionKey: string): string {
+    return `
+import * as ShellAPI from '/shell-api.ts';
+import * as bson from '/bson.ts';
+
+export type ServerSchema = ${this.connectionSchemas[
+      connectionKey
+    ].toTypescriptTypeDefinition()};
+`;
+  }
+
+  getCurrentGlobalsCode(connectionId: string, databaseName: string) {
+    return `
+import { ServerSchema } from '/${connectionId}.ts';
+
+type CurrentDatabaseSchema = ServerSchema['${databaseName}'];
+
+declare global {
+  var db: CurrentDatabaseSchema;
+  var use: (collection: string) => CurrentDatabaseSchema;
+}
+`;
+  }
+
+  async autocomplete(
+    code: string,
+    { connectionId, databaseName, position }: AutocompleteOptions
+  ): Promise<AutoCompletion[]> {
+    if (typeof position === 'undefined') {
+      position = code.length;
+    }
+
+    // TODO: we're now compiling the source twice: autocomplete will also
+    // compile if it finds completions in order to filter them. Does it matter?
+    // Maybe. Not so much for a typical short shell line, probably more for a
+    // large document or aggregation.
+    const tsAst = compileSourceFile(code);
+    const collectionName = inferCollectionNameFromFunctionCall(tsAst) || 'test';
+
+    const schema: JSONSchema =
+      await this.context.schemaInformationForCollection(
+        connectionId,
+        databaseName,
+        collectionName
+      );
+
+    const connection = this.addConnection(connectionId);
+    connection.addCollectionSchema(databaseName, collectionName, schema);
+
+    const collectionNames = await this.context.collectionsForDatabase(
+      connectionId,
+      databaseName
+    );
+    connection.setDatabaseCollectionNames(databaseName, collectionNames);
+
+    // TODO: the problem with doing it this way is that, while the collection
+    // schema might be cached, we'll be generating TypeScript for it (and every
+    // other collection in the db and in fact every db in the server) every
+    // time.
+    this.autocompleter.updateCode({
+      [`/${connectionId}.ts`]: this.getConnectionCode(connectionId),
+    });
+    this.autocompleter.updateCode({
+      '/current-globals.ts': this.getCurrentGlobalsCode(
+        connectionId,
+        databaseName
+      ),
```

Comment:
We can probably cache those quite easily? I agree that it might not be a problem, but it should be fairly low effort since we already have the caches.